### DISCUSSION TITLE
Optimized Docker image for Next.js + Custom server

### DISCUSSION THREAD

**Discussion by anton-johansson** [vercel/next.js#22149] [Help] [answered]
Hi!

I'm well aware of how to create an optimized Docker image for regular Next.js. By optimized, I mean to exclude everything that is not needed for the application to run, from the image. An example of that is `node_modules/`. Since stuff is compiled (into `.next/`?), we no longer need that in our production image. There is a good explaination of this here:
https://dev.to/chrsgrrtt/dockerising-a-next-js-project-1ck5

All we need is the `public/` directory, the compiled `.next/` directory and the `next` package to actually run it.

Now, when it comes to TypeScript, the example generated by `create-next-app` (`custom-server-typescript`) uses `next build` normally, but also `tsc` to compile the server directory into `dist/`. Then, to actually run the application, it simply does `node dist/index.js`.

Now, this works as long as I'm in the project directory. If I move `dist/` and `.next/` to a separate directory and run it, it'll complain about the `pages/` directory missing and such. It kind of feels like the code isn't "compiled" in the same smart manner as it is in the non-TS example.

Has anyone any good ideas on how to package the application in an optimal way when using TypeScript?

Thanks in advance!

**EDIT:**
After checking around in the samples, this is more related to "custom-server". All those examples seems to run `node` directly instead of using `next start`.

Still an issue, but it's probably not related to using TypeScript. :) Title updated.

Looking here:
https://nextjs.org/docs/advanced-features/custom-server

... you can pass in some parameters when creating the `next` app, such as `dir` (which defaults to `.`). Would it be possible to configure it to use a pre-compiled `.next` directory somehow?

**EDIT 2:**
Ok, after scouring through Next.js sources and debugging stuff, I realized that I did not have `NODE_ENV=production`, obviously causing Next.js to look for local sources... Sigh. Will update in a sec and close discussion if needed.

**Comment by anton-johansson** [comment, marked as answer]
Ok, so after struggling a bit, this is the final `Dockerfile` I came up with:

```Dockerfile
FROM node:14.15.0-alpine3.12 AS build
WORKDIR /usr/src

# Only build dependencies in the first step. This is very good for caching in Docker, as this layer will only
# be re-built whenever the dependency-list changes.
COPY package*.json ./
RUN npm ci

# Copy the actual source code and build it
COPY . .
ENV NODE_ENV=production
RUN npm run build

# Sadly, we need to include node_modules/ beacuse the server side needs it, but at least we can prune it to
# get rid of development dependencies.
RUN npm prune --production

# Start from scratch and include only relevant files
FROM node:14.15.0-alpine3.12 AS distribution
WORKDIR /opt/my-app
ENV NODE_ENV=production
COPY --from=build /usr/src/node_modules node_modules
COPY --from=build /usr/src/dist dist
COPY --from=build /usr/src/.next .next

# Expose port and run application
EXPOSE 3000
CMD node dist/index.js /etc/my-app/configuration.yaml
```

As you can see in one of the comments there, we sadly need to include `node_modules/` because the server side code needs it. It's very annoying, because that means that code is bundled twice (transpiled with `next build` for the client stuff) and through `node_modules/` for the server. However, it's probably an unrelated question, so I'm gonna close this discussion already.

**Reply by nazywamsiepawel** [reply]
Hey! You seem to have posted the only example of a Docker file for Next.js app with a custom server that I could find.

I'm trying to achieve a similar effect at the moment and I seem to be running into an issue with Next.js - "Couldn't find a `pages` directory with Next.js" - did you ever experience a similar one? More info on this StackOverflow thread - [https://stackoverflow.com/questions/70236426/couldnt-find-a-pages-directory-with-next-js-and-customer-server-build](https://stackoverflow.com/questions/70236426/couldnt-find-a-pages-directory-with-next-js-and-customer-server-build)

**Reply by anton-johansson** [reply]
Hey @nazywamsiepawel!

Oof, this was a while ago. Yeah, that is probably the exact same issue I was running into. As soon as I started building in production mode (`ENV NODE_ENV=production`), it started working as I expected it to. As far as I'm concerned, the pages will also be compiled into the distribution like everything else when you use production mode.

Are you using the environment variable mentioned above?
